v3/newrelic/sqlparse/sqlparse.go

```diff
@@ -4,47 +4,168 @@
 	if idx := strings.Index(s, "."); idx > 0 {
 		s = s[idx+1:]
 	}
-	return s
+
+	var buffer bytes.Buffer
+
+	for _, c := range s {
+		if unicode.IsSpace(c) || strings.ContainsRune("`'(){}[]\"", c) {
+			continue
+		} else {
+			buffer.WriteRune(c)
+		}
+	}
+
+	return buffer.String()
 }
 
-var (
-	basicTable        = `[^)(\]\[\}\{\s,;]+`
-	enclosedTable     = `[\[\(\{]` + `\s*` + basicTable + `\s*` + `[\]\)\}]`
-	tablePattern      = `(` + `\s+` + basicTable + `|` + `\s*` + enclosedTable + `)`
-	extractTableRegex = regexp.MustCompile(`[\s` + "`" + `"'\(\)\{\}\[\]]*`)
-	updateRegex       = regexp.MustCompile(`(?is)^update(?:\s+(?:low_priority|ignore|or|rollback|abort|replace|fail|only))*` + tablePattern)
-	sqlOperations     = map[string]*regexp.Regexp{
-		"select":   regexp.MustCompile(`(?is)^.*?\sfrom` + tablePattern),
-		"delete":   regexp.MustCompile(`(?is)^.*?\sfrom` + tablePattern),
-		"insert":   regexp.MustCompile(`(?is)^.*?\sinto?` + tablePattern),
-		"update":   updateRegex,
-		"call":     nil,
-		"create":   nil,
-		"drop":     nil,
-		"show":     nil,
-		"set":      nil,
-		"exec":     nil,
-		"execute":  nil,
-		"alter":    nil,
-		"commit":   nil,
-		"rollback": nil,
+// Returns a string slice of query without trailing spaces.
+func skipSpace(query string) string {
+	for i, c := range query {
+		if !unicode.IsSpace(c) {
+			return query[i:]
+		}
 	}
-	firstWordRegex   = regexp.MustCompile(`^\w+`)
-	cCommentRegex    = regexp.MustCompile(`(?is)/\*.*?\*/`)
-	lineCommentRegex = regexp.MustCompile(`(?im)(?:--|#).*?$`)
-	sqlPrefixRegex   = regexp.MustCompile(`^[\s;]*`)
-)
+	return query
+}
+
+// Returns a string slice of query with trailing comments removed.
+func skipComment(query string) string {
+	query = skipSpace(query)
+
+	if strings.HasPrefix(query, "/*") {
+		if commentEnd := strings.Index(query[2:], "*/"); commentEnd != -1 {
+			return skipComment(query[commentEnd+4:])
+		}
+		return ""
+	} else if strings.HasPrefix(query, "--") {
+		if commentEnd := strings.Index(query[2:], "\n"); commentEnd != -1 {
+			return skipComment(query[commentEnd+3:])
+		}
+		return ""
+	} else if strings.HasPrefix(query, ";") || strings.HasPrefix(query, "#") {
+		return skipComment(query[1:])
+	}
+
+	return query
+}
+
+// Returns a new string with all comments removed.
+func removeAllComments(query string) string {
+	query = skipComment(query)
+	for i, c := range query {
+		if strings.HasPrefix(query[i:], "/*") || strings.HasPrefix(query[i:], "--") || c == ';' || c == '#' {
+			query = fmt.Sprintf("%s%s", query[:i], skipComment(query[i:]))
+			return removeAllComments(query)
+		}
+	}
+	return query
+}
+
+// A SQL tokenizer that extracts tokens from an SQL string.
+type sqlTokenizer struct {
+	query string
+}
+
+// Returns the first word and the remainder of the given string (with trailing
+// comments removed).
+//
+// A word is either sequence of letters and digits, or a single special
+// character.
+func (q *sqlTokenizer) nextWord() string {
```

Comment:
I haven't thought this through all the way, but is there a reason why we get tokens "manually" here instead of using strings.Fields?

Reply:
Because SQL is nasty. This, for example, is a valid SQL query:
```sql
select*from table_name
```
And `strings.Fields` only splits at spaces.